zstd: add zstd package

ammubhave opened the pull request:
Zstandard, or zstd as short version, is a fast lossless compression
algorithm, targeting real-time compression scenarios at zlib-level and
better compression ratios.

This package includes a C library, and a command line utility producing
and decoding .zst, .gz, .xz and .lz4 files.

Maintainer: me
Compile tested: openwrt master, ipq806x
Run tested: openwrt master, ipq806x

diizzyy commented:
Please split library and userland utilities into separate packages
Add option to enforce O3 as that seems what upstream "recommends".
Source: https://github.com/facebook/zstd/releases

ammubhave commented:
UPDATE:
- Split into two packages zstd and libzstd, where the former provides zstd/unzstd binaries, and the latter provides the shared library.
- Add option to use O3 optimization.

diizzyy commented:
@ammubhave 
It's enforcing O3, not making it optional
Have a look here instead: https://github.com/openwrt/packages/tree/master/sound/fdk-aac

ammubhave commented:
Added an option CONFIG_ZSTD_OPTIMIZE_O3 to control -O3 optimization.

neheb commented:
Any reason for the libstdcpp dependency?

ammubhave commented:
Updated from version 1.3.5 -> 1.3.7

Removed the libstdcpp dependency, seems to compile fine with ulibc++.
There is a host build `$(eval $(call HostBuild))` specified at the very last line. This ensures packages for host tools can use the lib and zstd binary.

@diizzyy @neheb Let me know if you have any more suggestions or does it look good?
The package has a CONFIG_ZSTD_OPTIMIZE_O3 config to optimize O3 enabled by default.

neheb commented:
I tested this with btrfs-progs. So far so good.

I will make this a dependency in my next version bump.

diizzyy commented:
Doesn't that pass both -Os or whatever you have as default in OpenWrt's toolchain and -O3 additionally?

ammubhave commented:
I rechecked the optimize options, it seems "If you use multiple -O options, with or without level numbers, the last such option is the one that is effective." (ref: https://gcc.gnu.org/onlinedocs/gcc-4.3.4/gcc/Optimize-Options.html)

With this in mind, I added the optimize option to EXTRA_CXXFLAGS to ensure the -O option is at the end. Checked output of `ps` when running `make package/zstd/compile`, CXXFLAGS cmake variable shows O3 at the end of the list.

diizzyy commented:
Correct but it's ugly and potentially confusing, just strip it instead.

ammubhave commented:
Makes sense. Stripped out any `-O` from flags before appending O3.
Tested by checking output of ps and verified that `-Os` no longer shows up in flags.

neheb commented:
@Cynerd do you want this for btrfs-progs?

sedlund commented:
Good decompression speed increase in zstd 1.3.8 can you bump?

Also, can this get merged?

neheb commented:
@sedlund I am looking forward to do so. I added support for it to btrfs-progs when I used/maintained it.

diizzyy commented:
Given recent discussions, do we have any users of zstd apart from btrfs?

sedlund commented:
I am mainly interested in the zstd command line utility.  It would be the fastest compression program available in OpenWRT.  It including LZ4 is also a huge bonus for our low power devices.

neheb commented:
@diizzy libarchive is another. GNU tar as well. There are a few more.

neheb commented:
After thinking about this some, I've decided to merge. Adding support in other packages can happen later.
